### PR TITLE
Add trigger from building improvements and trigger conditional for building improvements

### PR DESCRIPTION
**Pull request by SeventhM**

### CODE REVIEW

**Comment by SeventhM** [comment]
Not sure if the wording should be changed to "upon building a [improvementFilter] improvement"

File: core/src/com/unciv/models/ruleset/unique/UniqueType.kt [hunk 1]
```diff
@@ -756,6 +756,7 @@ enum class UniqueType(
     /** Can be placed upon both units and as global */
     TriggerUponConqueringCity("upon conquering a city", UniqueTarget.TriggerCondition, UniqueTarget.UnitTriggerCondition),
     TriggerUponFoundingCity("upon founding a city", UniqueTarget.TriggerCondition),
+    TriggerUponBuildingImprovement("upon building [improvementFilter]", UniqueTarget.TriggerCondition, UniqueTarget.UnitTriggerCondition),
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/UniqueType.kt:759]: Definitely change this to "upon building a [improvementFilter] improvement", looks too much like "upon constructing" below and could easily confuse both players and modders

File: core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt [hunk 2]
```diff
@@ -223,6 +225,26 @@ class TileInfoImprovementFunctions(val tile: Tile) {
         )
             takeOverTilesAround(civToActivateBroaderEffects, tile)
 
+        if (civToActivateBroaderEffects != null && improvementObject != null) {
+            for (unique in improvementObject.uniqueObjects)
+                if (!unique.hasTriggerConditional()) {
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt:230]: Please add this 'if' as .uniqueObjects.filter{} instead, this decision tree is too long

File: core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt [hunk 3]
```diff
@@ -223,6 +225,26 @@ class TileInfoImprovementFunctions(val tile: Tile) {
...
+                }
+            if (unit != null){
+                for (unique in unit.getTriggeredUniques(UniqueType.TriggerUponBuildingImprovement)
+                    .filter { improvementObject.matchesFilter(it.params[0]) })
+                    UniqueTriggerActivation.triggerUnitwideUnique(unique, unit)
+            }
+            for (unique in civToActivateBroaderEffects.getMatchingUniques(
+                UniqueType.TriggerUponBuildingImprovement, StateForConditionals(civInfo = civToActivateBroaderEffects, unit = unit))
+                .filter { improvementObject.matchesFilter(it.params[0]) })
+                    UniqueTriggerActivation.triggerCivwideUnique(unique, civToActivateBroaderEffects)
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt:244]: In fact, ALL of this should be extracted to a different function

File: core/src/com/unciv/models/ruleset/unique/UniqueType.kt [hunk 4]
```diff
@@ -756,6 +756,7 @@ enum class UniqueType(
     /** Can be placed upon both units and as global */
     TriggerUponConqueringCity("upon conquering a city", UniqueTarget.TriggerCondition, UniqueTarget.UnitTriggerCondition),
     TriggerUponFoundingCity("upon founding a city", UniqueTarget.TriggerCondition),
+    TriggerUponBuildingImprovement("upon building a [improvementFilter] improvements", UniqueTarget.TriggerCondition, UniqueTarget.UnitTriggerCondition),
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/UniqueType.kt:759]:
```suggestion
    TriggerUponBuildingImprovement("upon building a [improvementFilter] improvement", UniqueTarget.TriggerCondition, UniqueTarget.UnitTriggerCondition),
```

**Pull request merged**